lib/boxcars.rb

```diff
@@ -61,25 +61,28 @@ def serpapi_api_key(**kwargs)
     private
 
     def check_key(key, val)
-      return val unless val.nil? || val.empty?
+      unless val.nil? || val.empty?
+        instance_variable_set("@#{key}", val)
+        return val
+      end
 
       error_text = ":#{key} missing! Please pass key, or set #{key.to_s.upcase} environment variable."
       raise ConfigurationError, error_text
     end
 
     def key_lookup(key, kwargs)
-      rv = if kwargs.key?(key) && !kwargs[key].nil?
-             # override with kwargs if present
-             kwargs[key]
-           elsif (set_val = instance_variable_get("@#{key}"))
-             # use saved value if present
-             set_val
-           else
-             # otherwise, dig out of the environment
-             new_key = ENV.fetch(key.to_s.upcase, nil)
-             new_key
-           end
-      check_key(key, rv)
+      val = if kwargs.key?(key) && !kwargs[key].nil?
+              # override with kwargs if present
+              kwargs[key]
+            elsif (saved_val = instance_variable_get("@#{key}"))
```

Comment:
I've noticed that here you were checking for the instance variable but did not find where do you create them, so I added those lines

Reply:
Here is how I thought it should work:

1) if you pass in the_key, use that for this instance.
2) if you configure Boxcars::configuration.the_key and it isn't passed in, use the configuration value. Note that this setting is not instance scope, but gem wide when using it.
3) if you have an ENV var set for the_key, and it isn't passed in, use the ENV var value.

In the code you added, I don't see where you create them, but I don't think we need to. Anyway, your code is clearer here.

Reply:
```
if you configure Boxcars::configuration.the_key and it isn't passed in, use the configuration value. Note that this setting is not instance scope, but gem wide when using it.
```

Meaning if you will do `Boxcars::configuration.the_key = "abcde"` ?

Reply:
yes, in this case `Boxcars::configuration.openai_access_token = "abcde"` or `Boxcars::configuration.serpapi_api_key = "abcde"`